File: inspection_report.vb
' Language: VBScript (CATScript)
' 検査自動化テンプレート
' 1) Publication名で対象形状を抽出
' 2) 断面生成（可能なら実行、失敗しても続行）
' 3) 寸法計測（例：最小距離）＋判定
' 4) 図面に「断面ラベル（簡易）＋判定表」を左右併記
' 5) PDF出力

Option Explicit

'========================
' 設定値（必要に応じて編集）
'========================
Const PUB_A_NAME = "U_TARGET"   ' Publication名（例：U側ターゲット）
Const PUB_B_NAME = "I_TARGET"   ' Publication名（例：I側ターゲット）
Const GAP_MIN_MM = 25.0         ' 合格下限（例：最小離隔 ≧ 25mm）

' 断面の基準面（OriginElementsのXY/YZ/ZXいずれかを使用）
Const SECTION_PLANE = "XY"      ' "XY" / "YZ" / "ZX"
Const DRAW_SHEET_FORMAT = "A4"
Const PDF_OUT_PATH = "C:\temp\inspection_report.pdf"  ' 出力先（存在する書き込み可ディレクトリ）

'========================
' メイン
'========================
Sub CATMain()
    On Error Resume Next

    If CATIA Is Nothing Then
        MsgBox "CATIAが起動していません。", vbCritical : Exit Sub
    End If

    If CATIA.Documents.Count = 0 Then
        MsgBox "アクティブなドキュメントがありません。Partを開いてから実行してください。", vbCritical : Exit Sub
    End If

    Dim doc : Set doc = CATIA.ActiveDocument
    If LCase(doc.Type) <> "part" Then
        MsgBox "PartDocument上で実行してください。現在: " & doc.Type, vbCritical : Exit Sub
    End If

    Dim partDoc : Set partDoc = doc
    Dim part    : Set part    = partDoc.Part

    '---- Publicationから対象形状を取得
    Dim refA : Set refA = GetRefFromPublication(part, PUB_A_NAME)
    Dim refB : Set refB = GetRefFromPublication(part, PUB_B_NAME)
    If (refA Is Nothing) Or (refB Is Nothing) Then
        MsgBox "Publicationが見つかりません。" & vbCrLf & _
               "A=" & PUB_A_NAME & ", B=" & PUB_B_NAME & vbCrLf & _
               "PartのPublicationに登録してください。", vbCritical
        Exit Sub
    End If

    '---- 断面（可能なら）作成：失敗しても計測に進む
    Dim sectionOk : sectionOk = CreateSectionCurveIfPossible(part, SECTION_PLANE)
    part.Update

    '---- 計測（例：A-Bの最小距離）
    Dim gapMM, p1, p2
    gapMM = 0 : Set p1 = Nothing : Set p2 = Nothing
    gapMM = GetMinimumDistance(part, refA, refB, p1, p2)

    '---- 判定
    Dim passFail, reason
    If gapMM >= GAP_MIN_MM Then
        passFail = "PASS"
        reason   = "最小離隔が閾値以上"
    Else
        passFail = "FAIL"
        reason   = "最小離隔が閾値未満"
    End If

    '---- 図面作成（左：断面ラベル／右：判定表）→ PDF出力
    Dim pdfOk : pdfOk = BuildDrawingAndExportPDF(partDoc, gapMM, passFail, reason, sectionOk, PDF_OUT_PATH)
    If pdfOk Then
        MsgBox "完了: " & PDF_OUT_PATH, vbInformation
    Else
        MsgBox "図面出力に失敗しました。保存先や権限を確認してください。", vbExclamation
    End If

End Sub

'========================
' Publication → Reference
'========================
Function GetRefFromPublication(part, pubName)
    On Error Resume Next
    Dim pubs : Set pubs = part.Publications
    Dim pub  : Set pub  = pubs.Item(pubName)
    If pub Is Nothing Then
        Set GetRefFromPublication = Nothing
        Exit Function
    End If
    Set GetRefFromPublication = pub.Valuation
End Function

'========================
' 最小距離（SPAWorkbench）
'========================
Function GetMinimumDistance(part, ref1, ref2, p1, p2)
    On Error Resume Next
    Dim spa : Set spa = part.Parent.GetWorkbench("SPAWorkbench")
    Dim meas1 : Set meas1 = spa.GetMeasurable(ref1)
    Dim meas2 : Set meas2 = spa.GetMeasurable(ref2)

    Dim d, x1, y1, z1, x2, y2, z2
    d = 0 : x1=0: y1=0: z1=0: x2=0: y2=0: z2=0

    ' GetMinimumDistanceは2つのMeasurable/Reference間の最近点距離と最近点座標を返す
    d = meas1.GetMinimumDistance(ref2, x1, y1, z1, x2, y2, z2)

    ' 返却用ポイントは使用しない場合もあるが、将来の寸法配置に活用可
    Set p1 = Array(x1, y1, z1)
    Set p2 = Array(x2, y2, z2)

    GetMinimumDistance = d
End Function

'========================
' 断面生成（可能なら実行）
' GSDのHybridShapeFactoryで原点面とBodyの交線を作成する例
' ※環境によりAPI差異があるため、失敗しても処理続行
'========================
Function CreateSectionCurveIfPossible(part, planeKey)
    On Error Resume Next

    Dim origin : Set origin = part.OriginElements
    Dim basePlane
    If UCase(planeKey) = "XY" Then
        Set basePlane = origin.PlaneXY
    ElseIf UCase(planeKey) = "YZ" Then
        Set basePlane = origin.PlaneYZ
    Else
        Set basePlane = origin.PlaneZX
    End If

    Dim hsf  : Set hsf  = part.HybridShapeFactory
    Dim hBodies : Set hBodies = part.HybridBodies
    Dim tgtBody
    ' 交線を置くジオメトリセット（なければ作成）
    Dim setName : setName = "CS_Section"
    Dim secSet
    On Error Resume Next
    Set secSet = hBodies.Item(setName)
    If secSet Is Nothing Then
        Set secSet = hBodies.Add()
        secSet.Name = setName
    End If

    ' 代表Body（最初のSolid）を対象に交線を作る簡易例
    ' ※製品構成や複数Bodyの場合は適宜拡張
    Dim bodies : Set bodies = part.Bodies
    If bodies.Count = 0 Then
        CreateSectionCurveIfPossible = False
        Exit Function
    End If
    Set tgtBody = bodies.Item(1)

    Dim refPlane : Set refPlane = part.CreateReferenceFromObject(basePlane)
    Dim refBody  : Set refBody  = part.CreateReferenceFromObject(tgtBody)

    ' 平面×ソリッドの交線（HybridShapeSection／Intersection系）
    Dim secFeat
    On Error Resume Next

    ' 1) Try: AddNewSection（環境により利用不可な場合あり）
    Set secFeat = hsf.AddNewSection(refBody, refPlane)
    If Err.Number <> 0 Or secFeat Is Nothing Then
        Err.Clear
        ' 2) Fallback: AddNewIntersection（こちらも環境差あり）
        Set secFeat = hsf.AddNewIntersection(refBody, refPlane)
    End If

    If Err.Number <> 0 Or secFeat Is Nothing Then
        Err.Clear
        CreateSectionCurveIfPossible = False
        Exit Function
    End If

    secSet.AppendHybridShape secFeat
    part.InWorkObject = secFeat
    part.Update

    CreateSectionCurveIfPossible = True
End Function

'========================
' 図面作成＋表記入＋PDF出力
' 左側：断面ラベル（簡易）
' 右側：判定表
'========================
Function BuildDrawingAndExportPDF(partDoc, gapMM, passFail, reason, sectionOk, pdfPath)
    On Error Resume Next

    Dim drw : Set drw = CATIA.Documents.Add("Drawing")
    If drw Is Nothing Then
        BuildDrawingAndExportPDF = False
        Exit Function
    End If

    Dim sheet : Set sheet = drw.Sheets.Item(1)
    sheet.PaperSize = DRAW_SHEET_FORMAT

    ' ビュー（左）作成
    Dim views : Set views = sheet.Views
    Dim vLeft : Set vLeft = views.Add("View_Left")
    vLeft.x = 30  ' 左余白
    vLeft.y = 180 ' 上からの位置（mm単位）
    vLeft.Scale = 1

    ' 断面ラベル（ここに断面ビュー生成を後で置換可能）
    Dim txts : Set txts = vLeft.Texts
    Dim sLabel
    If sectionOk Then
        sLabel = "断面: " & SECTION_PLANE & "（交線生成済）"
    Else
        sLabel = "断面: " & SECTION_PLANE & "（交線生成不可／環境未対応）"
    End If
    Call txts.Add(sLabel, 10, 140)
    Call txts.Add("寸法（例：A-B最小離隔）: " & FormatNumber(gapMM, 3) & " mm", 10, 125)

    ' ビュー（右）に表を作成
    Dim vRight : Set vRight = views.Add("View_Right")
    vRight.x = 140
    vRight.y = 180
    vRight.Scale = 1

    Dim tables : Set tables = sheet.Tables
    Dim tbl : Set tbl = tables.Add(135, 260, 4, 3, 35, 10) ' (X, Y, 行, 列, 列幅, 行高)

    ' ヘッダ
    Call tbl.SetCellString(1, 1, "検査項目")
    Call tbl.SetCellString(1, 2, "基準")
    Call tbl.SetCellString(1, 3, "結果")

    ' 行1：最小離隔
    Call tbl.SetCellString(2, 1, "最小離隔 A-B")
    Call tbl.SetCellString(2, 2, "≧ " & GAP_MIN_MM & " mm")
    Call tbl.SetCellString(2, 3, FormatNumber(gapMM, 3) & " mm")

    ' 行2：判定
    Call tbl.SetCellString(3, 1, "判定")
    Call tbl.SetCellString(3, 2, "—")
    Call tbl.SetCellString(3, 3, passFail)

    ' 行3：理由
    Call tbl.SetCellString(4, 1, "理由")
    Call tbl.SetCellString(4, 2, "—")
    Call tbl.SetCellString(4, 3, reason)

    ' 備考
    Dim tRightTxts : Set tRightTxts = vRight.Texts
    Call tRightTxts.Add("備考: Publication名 A=" & PUB_A_NAME & " / B=" & PUB_B_NAME, 135, 110)

    ' PDF出力
    drw.ExportData pdfPath, "pdf"

    BuildDrawingAndExportPDF = (Err.Number = 0)
End Function

File: inspection_report.py
# -*- coding: utf-8 -*-
"""
CATIA 検査自動化テンプレート（Python版 / COM）
- 3Dモデルから Publication 指定の形状を取得
- 可能なら断面交線を生成（失敗しても続行）
- A-B の最小距離を計測し、閾値で判定
- 図面上で「左：断面ラベル（簡易）／右：判定表」を左右併記
- PDF 出力

前提:
  - Windows + CATIA 環境
  - Python: pip install pywin32
  - CATIA を起動し、Part をアクティブにして実行
"""
from __future__ import annotations

import sys
from typing import Tuple, Optional

import win32com.client as win32
from win32com.client import constants


# ========================
# 設定値（必要に応じて編集）
# ========================
PUB_A_NAME = "U_TARGET"          # Publication名（A側）
PUB_B_NAME = "I_TARGET"          # Publication名（B側）
GAP_MIN_MM = 25.0                # 合格下限（最小離隔 ≧ この値）
SECTION_PLANE = "XY"             # "XY" / "YZ" / "ZX"
DRAW_SHEET_FORMAT = "A4"         # A0/A1/A2/A3/A4/A5
PDF_OUT_PATH = r"C:\temp\inspection_report.pdf"  # 出力先（存在する書き込み可ディレクトリ）


def get_catia():
    """起動中の CATIA.Application に接続（Dispatch）。"""
    try:
        return win32.Dispatch("CATIA.Application")
    except Exception as e:
        raise RuntimeError("CATIA に接続できません。CATIA を起動してください。") from e


def get_active_part_document(catia):
    """アクティブドキュメントが PartDocument かを確認して返す。"""
    if catia.Documents.Count == 0:
        raise RuntimeError("アクティブなドキュメントがありません。Part を開いてから実行してください。")
    doc = catia.ActiveDocument
    if str(doc.Type).lower() != "part":
        raise RuntimeError(f"PartDocument 上で実行してください。現在: {doc.Type}")
    return doc


def get_reference_from_publication(part, pub_name: str):
    """Publication 名から Reference を取得。見つからない場合は None。"""
    pubs = part.Publications
    try:
        pub = pubs.Item(pub_name)
    except Exception:
        return None
    return pub.Valuation  # Reference を返す


def get_origin_plane(part, plane_key: str):
    """原点要素から XY/YZ/ZX の平面オブジェクトを返す。"""
    origin = part.OriginElements
    key = plane_key.strip().upper()
    if key == "XY":
        return origin.PlaneXY
    if key == "YZ":
        return origin.PlaneYZ
    return origin.PlaneZX


def create_section_curve_if_possible(part, plane_key: str) -> bool:
    """
    断面交線をハイブリッドジオメトリセットに作成（可能な範囲）。
    - HybridShapeFactory.AddNewSection を試す → 失敗なら AddNewIntersection を試す。
    - いずれも環境差があるため、失敗しても False を返して続行可能。
    """
    try:
        base_plane = get_origin_plane(part, plane_key)
        hsf = part.HybridShapeFactory
        h_bodies = part.HybridBodies

        # ジオメトリ格納先のセット（なければ作成）
        set_name = "CS_Section"
        try:
            sec_set = h_bodies.Item(set_name)
        except Exception:
            sec_set = h_bodies.Add()
            sec_set.Name = set_name

        bodies = part.Bodies
        if bodies.Count == 0:
            return False
        tgt_body = bodies.Item(1)

        ref_plane = part.CreateReferenceFromObject(base_plane)
        ref_body = part.CreateReferenceFromObject(tgt_body)

        sec_feat = None
        # 1) AddNewSection を試す
        try:
            sec_feat = hsf.AddNewSection(ref_body, ref_plane)
        except Exception:
            sec_feat = None

        # 2) ダメなら AddNewIntersection を試す
        if sec_feat is None:
            try:
                # 引数順は実装により異なることがあるため、両順序を試す
                try:
                    sec_feat = hsf.AddNewIntersection(ref_body, ref_plane)
                except Exception:
                    sec_feat = hsf.AddNewIntersection(ref_plane, ref_body)
            except Exception:
                sec_feat = None

        if sec_feat is None:
            return False

        sec_set.AppendHybridShape(sec_feat)
        part.InWorkObject = sec_feat
        part.Update()
        return True

    except Exception:
        return False


def get_minimum_distance(part, ref1, ref2) -> Tuple[float, Tuple[float, float, float], Tuple[float, float, float]]:
    """
    A-B の最小距離と最近点座標を取得。
    pywin32 では GetMinimumDistance(ref) がタプルを返す実装が一般的:
      (distance, x1, y1, z1, x2, y2, z2)
    """
    spa = part.Parent.GetWorkbench("SPAWorkbench")
    meas1 = spa.GetMeasurable(ref1)

    # 距離と2点座標を一括取得
    distance, x1, y1, z1, x2, y2, z2 = meas1.GetMinimumDistance(ref2)
    p1 = (float(x1), float(y1), float(z1))
    p2 = (float(x2), float(y2), float(z2))
    return float(distance), p1, p2


def _paper_const(fmt: str):
    """紙サイズの定数マッピング（未定義は A4）。"""
    f = fmt.strip().upper()
    return {
        "A0": constants.catPaperA0,
        "A1": constants.catPaperA1,
        "A2": constants.catPaperA2,
        "A3": constants.catPaperA3,
        "A4": constants.catPaperA4,
        "A5": constants.catPaperA5,
    }.get(f, constants.catPaperA4)


def build_drawing_and_export_pdf(
    catia,
    part_doc,
    gap_mm: float,
    pass_fail: str,
    reason: str,
    section_ok: bool,
    pdf_path: str,
    sheet_format: str,
    section_plane: str,
    gap_min_mm: float,
) -> bool:
    """
    図面生成と PDF 出力。
    左ビュー: 断面ラベル（簡易）
    右ビュー: 判定表（4行×3列）
    """
    try:
        drw = catia.Documents.Add("Drawing")
        sheet = drw.Sheets.Item(1)
        sheet.PaperSize = _paper_const(sheet_format)

        # 左側ビュー
        views = sheet.Views
        v_left = views.Add("View_Left")
        v_left.x = 30.0     # 左余白
        v_left.y = 180.0    # 上からの位置（mm）
        v_left.Scale = 1.0

        txts = v_left.Texts
        s_label = (
            f"断面: {section_plane}（交線生成済）"
            if section_ok
            else f"断面: {section_plane}（交線生成不可／環境未対応）"
        )
        txts.Add(s_label, 10.0, 140.0)
        txts.Add(f"寸法（A-B最小離隔）: {gap_mm:.3f} mm", 10.0, 125.0)

        # 右側ビュー（判定表）
        v_right = views.Add("View_Right")
        v_right.x = 140.0
        v_right.y = 180.0
        v_right.Scale = 1.0

        tables = sheet.Tables
        # (X, Y, 行, 列, 列幅, 行高)
        tbl = tables.Add(135.0, 260.0, 4, 3, 35.0, 10.0)

        # ヘッダ
        tbl.SetCellString(1, 1, "検査項目")
        tbl.SetCellString(1, 2, "基準")
        tbl.SetCellString(1, 3, "結果")

        # 行1: 最小離隔
        tbl.SetCellString(2, 1, "最小離隔 A-B")
        tbl.SetCellString(2, 2, f"≧ {gap_min_mm:.3f} mm")
        tbl.SetCellString(2, 3, f"{gap_mm:.3f} mm")

        # 行2: 判定
        tbl.SetCellString(3, 1, "判定")
        tbl.SetCellString(3, 2, "—")
        tbl.SetCellString(3, 3, pass_fail)

        # 行3: 理由
        tbl.SetCellString(4, 1, "理由")
        tbl.SetCellString(4, 2, "—")
        tbl.SetCellString(4, 3, reason)

        # 備考
        v_right.Texts.Add(
            f"備考: Publication名 A={PUB_A_NAME} / B={PUB_B_NAME}",
            135.0,
            110.0,
        )

        # PDF 出力
        drw.ExportData(pdf_path, "pdf")
        return True
    except Exception:
        return False


def main():
    catia = get_catia()
    part_doc = get_active_part_document(catia)
    part = part_doc.Part

    # Publication から対象参照を取得
    ref_a = get_reference_from_publication(part, PUB_A_NAME)
    ref_b = get_reference_from_publication(part, PUB_B_NAME)
    if ref_a is None or ref_b is None:
        raise RuntimeError(
            f"Publication が見つかりません。A='{PUB_A_NAME}', B='{PUB_B_NAME}' を Part の Publications に登録してください。"
        )

    # 断面（可能なら）作成：失敗しても続行
    section_ok = create_section_curve_if_possible(part, SECTION_PLANE)
    part.Update()

    # 計測（最小距離）
    gap_mm, p1, p2 = get_minimum_distance(part, ref_a, ref_b)

    # 判定
    if gap_mm >= GAP_MIN_MM:
        pass_fail = "PASS"
        reason = "最小離隔が閾値以上"
    else:
        pass_fail = "FAIL"
        reason = "最小離隔が閾値未満"

    # 図面生成 + PDF 出力
    ok = build_drawing_and_export_pdf(
        catia=catia,
        part_doc=part_doc,
        gap_mm=gap_mm,
        pass_fail=pass_fail,
        reason=reason,
        section_ok=section_ok,
        pdf_path=PDF_OUT_PATH,
        sheet_format=DRAW_SHEET_FORMAT,
        section_plane=SECTION_PLANE,
        gap_min_mm=GAP_MIN_MM,
    )
    if not ok:
        raise RuntimeError("図面の作成または PDF 出力に失敗しました。保存先や権限を確認してください。")

    print(f"完了: {PDF_OUT_PATH}")
    print(f"最小離隔: {gap_mm:.3f} mm, 判定: {pass_fail}, 理由: {reason}")


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"[ERROR] {exc}", file=sys.stderr)
        sys.exit(1)
